### PR TITLE
Make cuda default stream wait for operations in other streams by default

### LINKED ISSUE #2702: The current CUDA stream is not honored by some operations
Hi, I think something is wrong with handling CUDA streams.  Could someone take a look?

(I'm using Python 2.7.12, CUDA 8.0, and PyTorch version 0.2.0.post1)

    #!/usr/bin/python
    # PyTorch CUDA stream test.

    import torch

    DIM = 2048
    M = torch.cuda.FloatTensor(DIM, DIM)

    def do_test(M, use_stream, fn):
        M.fill_(0.1)
        print 'Initial value = %.4f' % fn()

        # Create a new stream if requested.
        stream = torch.cuda.Stream() if use_stream else None
        with torch.cuda.stream(stream):
            for step in range(5):
                # Do random calculation.
                K = torch.matmul(M, M) / DIM
                M -= K
                print 'After step %d = %.4f' % (step, fn())

    fn = M.norm   # or 'M.sum' or 'lambda: M[0, 0]'

    print '=== Without using a new stream ==='
    do_test(M, False, fn)
    print '=== With using a new stream ==='
    do_test(M, True, fn)

Result on my machine (GTX 1080).  The second half keeps changing:

    === Without using a new stream ===
    Initial value = 204.8000
    After step 0 = 184.3197
    After step 1 = 167.7312
    After step 2 = 153.9943
    After step 3 = 142.4149
    After step 4 = 132.5118
    === With using a new stream ===
    Initial value = 204.8000
    After step 0 = 3.1469
    After step 1 = 3.1469
    After step 2 = 3.1469
    After step 3 = 3.1469
    After step 4 = 3.1469

Basically, when I'm using a different stream, computing any single float value returns random results.  Running nvprof shows that when the second do_test() runs, all the calculations happen in the new stream (13 below), but copying the result back to CPU happens in the default stream (7 below), and at the wrong moments before the correct value is ready.

    $ nvprof --print-gpu-trace ./cuda-stream-test.py

       Start  Duration  Size  Stream  Name
    ... (only showing the second half)
    2.28038s     320ns    4B       7  [CUDA memcpy DtoH]
    2.28086s  2.0956ms     -      13  sgemm_128x128x8_NN_vec [796]
    2.28132s     960ns    4B       7  [CUDA memcpy DtoH]
    2.28183s  1.1200us    4B       7  [CUDA memcpy DtoH]
    2.28194s     960ns    4B       7  [CUDA memcpy DtoH]
    2.28206s  1.0560us    4B       7  [CUDA memcpy DtoH]
    2.28218s  1.1200us    4B       7  [CUDA memcpy DtoH]
    2.28296s  147.40us     -      13  void kernelPointwiseApply2<TensorDivConstantOp<float>, float, float, unsigned int, int=-2, int=-2>(TensorInfo<TensorDivConstantOp<float>, float>, TensorInfo<float, float>, float, float) [814]
    2.28311s  212.52us     -      13  void kernelPointwiseApply2<TensorSubOp<float>, float, float, unsigned int, int=-2, int=-2>(TensorInfo<TensorSubOp<float>, float>, TensorInfo<float, float>, float, float) [827]
    2.28332s  93.890us     -      13  void kernelReduceAllPass1<TensorNormOp<float, int=2>, ReduceAdd<float, float>, ReduceAdd<float, float>, float, float, unsigned int, int=-2>(TensorInfo<float, ReduceAdd<float, float>>, ReduceAdd<float, float>, float, float, int=2, TensorNormOp<float, int=2>, float*) [857]
    2.28341s  2.8160us     -      13  void kernelReduceAllPass2<ReduceAdd<float, float>, float, unsigned int>(int, float, float, float*, float*) [866]
    2.28342s  2.0951ms     -      13  sgemm_128x128x8_NN_vec [892]
    ... (4 more calls to these kernels)

Please enlighten me if I'm doing something wrong.  Thanks!

### PR DESCRIPTION
**Pull request by zou3519**
### Summary
Fixes https://github.com/pytorch/pytorch/issues/2702.

The bug is that cuda streams are created with the `CudaStreamNonBlocking` flag which means that operations on default streams do not block to wait for operations on the created stream. Matrix operations are run on some non-default stream while device to host transfers run on the default stream, leading to some race conditions.

Another way to fix the issue is to change device-to-host transfers to run on the cuda current stream but I think removing the `CudaStreamNonBlocking` flag makes it easier to reason about the default stream vs other streams.

### Test Plan
Run the code provided in https://github.com/pytorch/pytorch/issues/2702. Assert that junk data is not printed out.

### CODE REVIEW

**Comment by colesbury** [comment]
The existing behavior is correct. If you use CUDA streams, you need to insert the appropriate synchronizations yourself.

**Comment by soumith** [comment]
yea, this is not a right fix. we rely on and need non-blocking streams where appropriate. you have to insert stream synchronizations yourself if you are on a non-default stream.

**Comment by colesbury** [comment]
Sorry -- I missed the important point in the issue when I first read it. To clarify a bit:

1) The `CudaStreamNonBlocking` is intentional and important. The model for PyTorch is that operations in **all** streams are asynchronous with respect to each and the host. This makes it easier to write performant code that doesn't have subtle synchronizations.

2) This means that every operation should happen in the "current stream. We have some memcpy's in the default stream that need to be fixed.

3) The MAGMA and RNG functions don't support streams for the most part. We can probably leave them as-is for now.

**Comment by zou3519** [comment]
Thanks for the clarification, that makes sense. I'll move the applicable memcpy's to use the default stream.

Should the memcpy's still be synchronous with respect to the host (ie, I would change each `cudaMemcpy` to a `cudaMemcpyAsync` + `cudaSynchronizeStream(stream)` call)? I'm thinking yes because it would be slightly awkward for someone to do the following:
```
x = x.cuda()  # some tensor
some_tensor_op(x)
print(x)  # copies device to host
```
and realize they need to synchronize the stream they're using manually before right before the print statement.

**Comment by colesbury** [comment]
Yes, they should be synchronous with respect to the host
